### ISSUE TITLE
Column Backfill causing ERROR: command is too large: 268046651 bytes (max: 67108864)

### ISSUE THREAD

**Issue by daniel-crlabs** [cockroachdb/cockroach#63634]
**Describe the problem**

During an `ALTER TABLE` operation, adding a computed column to a 6.3 Terabyte table fails with command too large error:

```
ALTER TABLE aggregation_history
ADD COLUMN ehid string as ( md5(concat(entity_id, entity_type)) ) STORED;
ERROR: command is too large: 268046651 bytes (max: 67108864)
```

**To Reproduce**

Attempt to run the following alter statement:

```
ALTER TABLE aggregation_history
ADD COLUMN ehid string as ( md5(concat(entity_id, entity_type)) ) STORED;
```

**Additional context**

This costumer has experienced this same problem in the past, although with a different operation. They were provided with a custom binary, however, since this is for a higher limit, this is no longer working.

In the past, they have also attempted to increase the `kv.raft.command.max_size` but it caused the cluster to crash with a panic loop, so increasing the limit is not an option at the moment.

I have found a [github issue (61822)](https://github.com/cockroachdb/cockroach/issues/61822) somewhat addressing this, but it doesn't appear to work for them any longer, as the fix was for 128MB and the error now is for a larger size, 256MB. (???)

Further research shows this [PR (61818)](https://github.com/cockroachdb/cockroach/pull/61818) is in progress, to increase this to 512MB?

gz#8140

**Comment by knz** [comment]
I was asking:

> what is the maximum length of the existing columns entity_id  and entity_type?
> like, select max(entity_id), max(entitity_type) from aggregation_history
> it's also useful to see if these columns are part of a column family, and what is the max width of a row in these column families

Andrew:

>  I'm guessing that the rub here is that the backfill batches too many writes into a single request. In 20.2.7 I believe @Sumeer Bhola has a fix to allow kv.raft.command.max_size to go up to 512MB and avoid the crashes they got last time (on 20.2.6 I think). https://github.com/cockroachdb/cockroach/pull/61818

Andrei:

> maybe if they add the new column as its own col fam then the size of the rows doesn't matter?

Daniel:

> oh, and I just noticed this in the ticket, apparently this table is huuuuuuuuge. 6.3TiB table

**Comment by ajwerner** [comment]
The root cause here is that the column backfiller runs a chunk of rows of a constant count rather than a byte size.

https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/rowexec/backfiller.go#L145
The code that runs the column backfill should have some maximum memory size above which it also runs the batch:

https://github.com/cockroachdb/cockroach/blob/87ed071f5a23295b34fd4d56d4539f0d0207f937/pkg/sql/backfill/backfill.go#L319

**Comment by daniel-crlabs** [comment]
I had asked the customer for the above, however, I have not received a response yet. 
I'll update this issue whenever I hear back from them with that information.

**Comment by knz** [comment]
@daniel-crlabs it's useful to transcribe all the conversation in the github issue as it's happening, so that newcomers to the convo who weren't there at the beginning know what went on until now.

**Comment by daniel-crlabs** [comment]
Here's the latest update from the customer, as expected, he is unable to perform certain operations on this table, as it takes a long time (over 6.3TB size table). Details below as provided to us by the customer via the zendesk ticket:



>I can't run the query above as it does a full scan and never completes. I think the statistics and histograms below accomplish mostly the same...

```
root@revenue-platform-crdb.us-west-2.aws.ddnw.net:26257/revenue_platform_prod> show statistics for table aggregation_history;
statistics_name | column_names | created | row_count | distinct_count | null_count | histogram_id
------------------+--------------------------------------------------------+----------------------------------+-----------+----------------+------------+---------------------
__auto__ | {<unknown>} | 2021-04-14 20:55:10.058229+00:00 | 0 | 0 | 0 | NULL
__auto__ | {<unknown>,created_at} | 2021-04-14 20:55:10.058229+00:00 | 0 | 0 | 0 | NULL
__auto__ | {ehid,version} | 2021-04-16 03:19:15.31041+00:00 | 203440306 | 203344534 | 0 | NULL
__auto__ | {entity_id} | 2021-04-16 03:19:15.31041+00:00 | 203440306 | 28766085 | 0 | 650355524757815366
__auto__ | {aggregation_id,entity_id} | 2021-04-16 03:19:15.31041+00:00 | 203440306 | 39440753 | 0 | NULL
__auto__ | {entity_type} | 2021-04-16 03:19:15.31041+00:00 | 203440306 | 12 | 0 | 650355524876664902
__auto__ | {aggregation_id,entity_id,entity_type} | 2021-04-16 03:19:15.31041+00:00 | 203440306 | 39216035 | 0 | NULL
.....
.....

root@revenue-platform-crdb.us-west-2.aws.ddnw.net:26257/revenue_platform_prod> show histogram 650355524757815366;
upper_bound | range_rows | distinct_range_rows | equal_rows
-----------------------------------------+------------+---------------------+-------------
'002127b4-1143-46e4-8807-ad54e0c724ee' | 0 | 0 | 20344
'1352c62e-c04a-469b-bddc-e07459e460f3' | 996857 | 143829.42572141453 | 20344
'1573143587' | 996857 | 143829.42572141453 | 20344
'1586601260' | 996857 | 143829.42572141453 | 20344
'1596090786' | 996857 | 143829.42572141453 | 20344
'1604783592' | 996857 | 143829.42572141453 | 20344
'1615514524' | 996857 | 143829.42572141453 | 20344
'1624575670' | 996857 | 143829.42572141453 | 20344
'1630729407' | 996857 | 143829.42572141453 | 20344
'1637357691' | 996857 | 143829.42572141453 | 20344
'1637566942' | 996857 | 143829.42572141453 | 20344
'1637725317' | 996857 | 143829.42572141453 | 20344
'1637862688' | 996857 | 143829.42572141453 | 20344
'1637989256' | 996857 | 143829.42572141453 | 20344
'1638134767' | 996857 | 143829.42572141453 | 20344
'1638265475' | 996857 | 143829.42572141453 | 20344
'1638390229' | 996857 | 143829.42572141453 | 20344
'1638545222' | 996857 | 143829.42572141453 | 20344
'1638687128' | 996857 | 143829.42572141453 | 20344
'1638829815' | 996857 | 143829.42572141453 | 20344
'1638928893' | 996857 | 143829.42572141453 | 20344
'1639050360' | 996857 | 143829.42572141453 | 20344
'1639174664' | 996857 | 143829.42572141453 | 20344
'1639316736' | 996857 | 143829.42572141453 | 20344
'1639466044' | 996857 | 143829.42572141453 | 20344
'1639597712' | 996857 | 143829.42572141453 | 20344
'1639749092' | 996857 | 143829.42572141453 | 20344
'1639865737' | 996857 | 143829.42572141453 | 20344
'1639999776' | 996857 | 143829.42572141453 | 20344
'1640149143' | 996857 | 143829.42572141453 | 20344
'1640237104' | 996857 | 143829.42572141453 | 20344
'1640342905' | 996857 | 143829.42572141453 | 20344
'1640475055' | 996857 | 143829.42572141453 | 20344
'1640605350' | 996857 | 143829.42572141453 | 20344
'1640783616' | 996857 | 143829.42572141453 | 20344
'1640891112' | 996857 | 143829.42572141453 | 20344
'1640993631' | 996857 | 143829.42572141453 | 20344
'1641102522' | 996857 | 143829.42572141453 | 20344
'1641236742' | 996857 | 143829.42572141453 | 20344
'1641340489' | 996857 | 143829.42572141453 | 20344
'1641452689' | 996857 | 143829.42572141453 | 20344
'1641556172' | 996857 | 143829.42572141453 | 20344
'1641703909' | 996857 | 143829.42572141453 | 20344
'1641852172' | 996857 | 143829.42572141453 | 20344
'1641992653' | 996857 | 143829.42572141453 | 20344
'1642122283' | 996857 | 143829.42572141453 | 20344
'1642285439' | 996857 | 143829.42572141453 | 20344
'1642407375' | 996857 | 143829.42572141453 | 20344
'1642539644' | 996857 | 143829.42572141453 | 20344
'1642653602' | 996857 | 143829.42572141453 | 20344
'1642771797' | 996857 | 143829.42572141453 | 20344
'1642896347' | 996857 | 143829.42572141453 | 20344
'1643040370' | 996857 | 143829.42572141453 | 20344
'1643183384' | 996857 | 143829.42572141453 | 20344
'1643311771' | 996857 | 143829.42572141453 | 20344
'1643435949' | 996857 | 143829.42572141453 | 20344
'2059d5f6-c47a-4ffd-8cdb-58a811c7f43a' | 996857 | 143829.42572141453 | 20344
'30375221' | 996857 | 143829.42572141453 | 20344
'419ac1d3-68d1-420c-9f77-8c293532f03c' | 996857 | 143829.42572141453 | 20344
'4373793867' | 996857 | 143829.42572141453 | 20344
'4379206806' | 996857 | 143829.42572141453 | 20344
'4384667441' | 996857 | 143829.42572141453 | 20344
'4388702171' | 996857 | 143829.42572141453 | 20344
'4393622340' | 996857 | 143829.42572141453 | 20344
'4399449873' | 996857 | 143829.42572141453 | 20344
'4404583598' | 996857 | 143829.42572141453 | 20344
'4410488647' | 996857 | 143829.42572141453 | 20344
'4414976688' | 996857 | 143829.42572141453 | 20344
'4418924358' | 996857 | 143829.42572141453 | 20344
'4422277486' | 996857 | 143829.42572141453 | 20344
'4428140921' | 996857 | 143829.42572141453 | 20344
'4432923285' | 996857 | 143829.42572141453 | 20344
'4438737088' | 996857 | 143829.42572141453 | 20344
'4442549937' | 996857 | 143829.42572141453 | 20344
'4446961378' | 996857 | 143829.42572141453 | 20344
'4448807692' | 996857 | 143829.42572141453 | 20344
'4449399412' | 996857 | 143829.42572141453 | 20344
'4449518010' | 996857 | 143829.42572141453 | 20344
'4449596450' | 996857 | 143829.42572141453 | 20344
'4449654695' | 996857 | 143829.42572141453 | 20344
'4449716327' | 996857 | 143829.42572141453 | 20344
'4449787925' | 996857 | 143829.42572141453 | 20344
'4449847463' | 996857 | 143829.42572141453 | 20344
'4449897044' | 996857 | 143829.42572141453 | 20344
'4449949679' | 996857 | 143829.42572141453 | 20344
'4450014427' | 996857 | 143829.42572141453 | 20344
'4450091493' | 996857 | 143829.42572141453 | 20344
'4450160646' | 996857 | 143829.42572141453 | 20344
'4450228722' | 996857 | 143829.42572141453 | 20344
'4450292408' | 996857 | 143829.42572141453 | 20344
'4450343770' | 996857 | 143829.42572141453 | 20344
'4450416541' | 996857 | 143829.42572141453 | 20344
'4450457167' | 996857 | 143829.42572141453 | 20344
'4450515028' | 996857 | 143829.42572141453 | 20344
'4450575911' | 996857 | 143829.42572141453 | 20344
'4450620507' | 996857 | 143829.42572141453 | 20344
'4450686742' | 996857 | 143829.42572141453 | 20344
'4450734178' | 996857 | 143829.42572141453 | 20344
'4450794610' | 996857 | 143829.42572141453 | 20344
'4450849502' | 996857 | 143829.42572141453 | 20344
'4450914057' | 996857 | 143829.42572141453 | 20344
'4450982264' | 996857 | 143829.42572141453 | 20344
'4451036831' | 996857 | 143829.42572141453 | 20344
'4451086926' | 996857 | 143829.42572141453 | 20344
'4451149169' | 996857 | 143829.42572141453 | 20344
'4451204839' | 996857 | 143829.42572141453 | 20344
'4451260653' | 996857 | 143829.42572141453 | 20344
'4451328402' | 996857 | 143829.42572141453 | 20344
'4451387980' | 996857 | 143829.42572141453 | 20344
'4451456036' | 996857 | 143829.42572141453 | 20344
'4451522891' | 996857 | 143829.42572141453 | 20344
'4451581988' | 996857 | 143829.42572141453 | 20344
'4451647507' | 996857 | 143829.42572141453 | 20344
'4451709471' | 996857 | 143829.42572141453 | 20344
'4451765784' | 996857 | 143829.42572141453 | 20344
'4451824254' | 996857 | 143829.42572141453 | 20344
'4451906854' | 996857 | 143829.42572141453 | 20344
'4451961334' | 996857 | 143829.42572141453 | 20344
'4452007216' | 996857 | 143829.42572141453 | 20344
'4452056085' | 996857 | 143829.42572141453 | 20344
'4452128648' | 996857 | 143829.42572141453 | 20344
'4452173680' | 996857 | 143829.42572141453 | 20344
'4452242203' | 996857 | 143829.42572141453 | 20344
'4452300808' | 996857 | 143829.42572141453 | 20344
'4452358661' | 996857 | 143829.42572141453 | 20344
'4452422210' | 996857 | 143829.42572141453 | 20344
'4452485798' | 996857 | 143829.42572141453 | 20344
'4452561989' | 996857 | 143829.42572141453 | 20344
'4452622572' | 996857 | 143829.42572141453 | 20344
'4452704321' | 996857 | 143829.42572141453 | 20344
'4452762226' | 996857 | 143829.42572141453 | 20344
'4452830490' | 996857 | 143829.42572141453 | 20344
'4452895003' | 996857 | 143829.42572141453 | 20344
'4452956035' | 996857 | 143829.42572141453 | 20344
'4453029374' | 996857 | 143829.42572141453 | 20344
'4453108170' | 996857 | 143829.42572141453 | 20344
'4453164059' | 996857 | 143829.42572141453 | 20344
'4453226837' | 996857 | 143829.42572141453 | 20344
'4453307205' | 996857 | 143829.42572141453 | 20344
'4453359925' | 996857 | 143829.42572141453 | 20344
'4453423464' | 996857 | 143829.42572141453 | 20344
'4453486050' | 996857 | 143829.42572141453 | 20344
'4453548482' | 996857 | 143829.42572141453 | 20344
'4453605215' | 996857 | 143829.42572141453 | 20344
'4453666727' | 996857 | 143829.42572141453 | 20344
'4453737569' | 996857 | 143829.42572141453 | 20344
'4453795967' | 996857 | 143829.42572141453 | 20344
'4453858974' | 996857 | 143829.42572141453 | 20344
'4453926826' | 996857 | 143829.42572141453 | 20344
'4453978591' | 996857 | 143829.42572141453 | 20344
'4454041325' | 996857 | 143829.42572141453 | 20344
'4454098079' | 1017201 | 146764.71717934325 | 20344
'4454172092' | 1017201 | 146764.71717934325 | 20344
'4454246301' | 1017201 | 146764.71717934325 | 20344
'4454314094' | 1017201 | 146764.71717934325 | 20344
'4454367513' | 1017201 | 146764.71717934325 | 20344
'4454429799' | 1017201 | 146764.71717934325 | 20344
'4454489567' | 1017201 | 146764.71717934325 | 20344
'4454549960' | 1017201 | 146764.71717934325 | 20344
'4454618587' | 1017201 | 146764.71717934325 | 20344
'4454672386' | 1017201 | 146764.71717934325 | 20344
'4454731594' | 1017201 | 146764.71717934325 | 20344
'4454786689' | 1017201 | 146764.71717934325 | 20344
'4454845158' | 1017201 | 146764.71717934325 | 20344
'4454925487' | 1017201 | 146764.71717934325 | 20344
'4454997333' | 1017201 | 146764.71717934325 | 20344
'4455047224' | 1017201 | 146764.71717934325 | 20344
'4455119282' | 1017201 | 146764.71717934325 | 20344
'4455190297' | 1017201 | 146764.71717934325 | 20344
'4455256198' | 1017201 | 146764.71717934325 | 20344
'4455319061' | 1017201 | 146764.71717934325 | 20344
'4455380910' | 1017201 | 146764.71717934325 | 20344
'4455436814' | 1017201 | 146764.71717934325 | 20344
'4455505282' | 1017201 | 146764.71717934325 | 20344
'4455576077' | 1017201 | 146764.71717934325 | 20344
'4455643446' | 1017201 | 146764.71717934325 | 20344
'4455715612' | 1017201 | 146764.71717934325 | 20344
'4455794720' | 1017201 | 146764.71717934325 | 20344
'4455877775' | 1017201 | 146764.71717934325 | 20344
'47cfee77-a4d7-42d0-807f-5f2efb672281' | 1017201 | 146764.71717934325 | 20344
'51468291' | 1017201 | 146764.71717934325 | 20344
'5d9a0f37-6b83-4087-842c-149de165e777' | 1017201 | 146764.71717934325 | 20344
'729809370' | 1017201 | 146764.71717934325 | 20344
'743701783' | 1017201 | 146764.71717934325 | 20344
'748665638' | 1017201 | 146764.71717934325 | 20344
'751251325' | 1017201 | 146764.71717934325 | 20344
'751516271' | 1017201 | 146764.71717934325 | 20344
'751762674' | 1017201 | 146764.71717934325 | 20344
'752071427' | 1017201 | 146764.71717934325 | 20344
'752293135' | 1017201 | 146764.71717934325 | 20344
'752503233' | 1017201 | 146764.71717934325 | 20344
'752688324' | 1017201 | 146764.71717934325 | 20344
'752972951' | 1017201 | 146764.71717934325 | 20344
'7db947d7-7e0b-4b6f-8bfe-a900a6fb6d87' | 1017201 | 146764.71717934325 | 20344
'9316d0e3-455c-417c-b173-6a620996e276' | 1017201 | 146764.71717934325 | 20344
'a2dacf7e-b237-4a8b-9e23-3ad24cb22d30' | 1017201 | 146764.71717934325 | 20344
'ba91150a-c18e-4620-a814-b8e120d6a5c0' | 1017201 | 146764.71717934325 | 20344
'd03e1fbf-ac74-4899-99db-41d6aa467462' | 1017201 | 146764.71717934325 | 20344
'eb0490fc-6264-4caf-a145-4031f9f5a99e' | 1017201 | 146764.71717934325 | 20344
'ff8bb817-5221-4b2f-be3e-76c9dacf5a70' | 1017201 | 146764.71717934325 | 20344
(200 rows)

root@revenue-platform-crdb.us-west-2.aws.ddnw.net:26257/revenue_platform_prod> show histogram 650355524876664902;
upper_bound | range_rows | distinct_range_rows | equal_rows
---------------------+------------+---------------------+-------------
'cart_payment' | 0 | 0 | 11331625
'credit_unit' | 488256 | 2.142857142857143 | 1586834
'dasher_shift' | 142408 | 0.6249999999999999 | 10965432
'delivery' | 0 | 0 | 54888194
'order' | 488256 | 2.142857142857143 | 6347337
'order_cart' | 0 | 0 | 90937816
'store_order_cart' | 20344 | 0.08928571428571427 | 26243799
```

DDL....

```
root@revenue-platform-crdb.us-west-2.aws.ddnw.net:26257/revenue_platform_prod> show create aggregation_history;
table_name | create_statement
-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------
aggregation_history | CREATE TABLE public.aggregation (
| faux_id INT8 NOT NULL DEFAULT unique_rowid(),
| aggregation_id UUID NOT NULL,
| entity_type STRING NOT NULL,
| entity_id STRING NOT NULL,
| created_at TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
| data JSONB NULL,
| version DECIMAL NULL DEFAULT 0:::DECIMAL,
| change_type STRING NULL,
| change_reference STRING NULL,
| CONSTRAINT "primary" PRIMARY KEY (aggregation_id ASC, entity_id ASC, entity_type ASC, created_at ASC, faux_id ASC),
| FAMILY "primary" (faux_id, aggregation_id, entity_type, entity_id, created_at, data, version, change_type, change_reference)
| );
```

**Comment by glennfawcett** [comment]
This was needed to work-around a computed column / index build issue. To work around this issue I a created another table with the index in-place and moved our traffic to the new table. I then pointed tried to copy data from the old table to the new table with a writeable CTE... but got this SAME error.

```
cockroach sql --certs-dir=./  --url postgresql://mydb:26257/ --database revenue_platform_prod --execute """
set disallow_full_table_scans=false;
WITH copybatch AS (
    DELETE FROM old_table
        WHERE 1=1
    LIMIT 400
            RETURNING old_table.*
    )
    INSERT INTO new_table
    SELECT DISTINCT * FROM copybatch;
    """ --watch 0.01s
ERROR: command is too large: 404137617 bytes (max: 67108864)
Failed running "sql"
```
Need to have a method to bulk move data from within the cluster in a fashion where the DB adjusts the size dynamically until the operation is complete. I wrote a script to parallelize the is effort, but indeed this should be able to be controlled from within the cluster.

**Comment by ajwerner** [comment]
@dt is there more we want to do here?

**Comment by dt** [comment]
I think #63936 gets us about as good as it'll get -- while the default was already quite small (and probably impacted performance for how small it was, at least until the deferred commit waits), with the knob you could in theory go all the way down to one-row-at-a-time. Beyond that, I don't think there's much for us to do here, since there are limits on size of a given row or single write already.

**Comment by ajwerner** [comment]
In that case, I'm closing this.

**Comment by daniel-crlabs** [comment]
Thanks for the updates. What do I tell the customer? Is this fixed on 63936  and if so, what release can they expect this to be available? Client is on 20.2.7. Please let me know and I'll relay that information to the client.
